### PR TITLE
Implement txn in invitation service

### PR DESCRIPTION
**Pull request by Junoth**
Tested in Insomnia and also added unit tests
Screen Shot 2023-04-02 at 3 34 45 PM
Screen Shot 2023-04-02 at 3 38 41 PM
Screen Shot 2023-04-02 at 3 39 23 PM
Screen Shot 2023-04-02 at 3 39 47 PM
Screen Shot 2023-04-02 at 3 40 21 PM
Screen Shot 2023-04-02 at 3 40 30 PM

(task:13707)

### CODE REVIEW

File: core/daov2/daotestv2/invitation.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+				if currInvitation.ID == invitationID {
+					invitation = currInvitation
+					return nil
+				}
+			}
+
+			return &errs.Error{
+				Code:    errs.NotFound,
+				Message: fmt.Sprintf("row not found: invitationID=%v", invitationID),
+			}
```
Comment (naturetime10) [inline core/daov2/daotestv2/invitation.go]: Let's use `NewError`

File: core/daov2/sqldb/invitation.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+			&invitation.ExpireAt,
+			&invitation.Status,
+			&invitation.Code,
+			&invitation.CreatedAt,
+			&invitation.UpdatedAt,
+		)
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return entity.Invitation{}, errs.NewError(errs.NotFound, fmt.Sprintf("invitation not found: invitationID=%v", invitationID))
```
Comment (naturetime10) [inline core/daov2/sqldb/invitation.go]:
```suggestion
			return entity.Invitation{}, errs.NewError(
			    errs.NotFound, 
			    fmt.Sprintf("invitation not found: invitationID=%v", invitationID))
```

File: core/service/invitation.go [hunk 3]
```diff
@@ -75,37 +82,25 @@ func (i Invitation) FindInvitations(ct context.Context, filter *InvitationFilter
...
 		}
 
 		if !hasPermission {
-			internalErr := &errs.Error{
-				Code:    errs.PermissionDenied,
-				Message: fmt.Sprintf("permission denied: authorization query=%v", query),
-			}
-			i.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-			return entity.Invitation{}, internalErr
+			return entity.Invitation{}, errs.NewError(errs.PermissionDenied, fmt.Sprintf("permission denied: authorization query=%v", query))
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
			return entity.Invitation{}, errs.NewError(
			    errs.PermissionDenied, 
			    fmt.Sprintf("permission denied: authorization query=%v", query))
```

File: core/service/invitation.go [hunk 4]
```diff
@@ -225,73 +243,134 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+		query := authorization.NewTeamAddMemberToQuery(receiverUserID, invitation.TeamID)
+		hasPermission, internalErr := i.authorizer.hasPermission(ct, query)
+		if internalErr != nil {
+			return entity.Invitation{}, err
 		}
-		i.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return entity.Invitation{}, internalErr
+
+		if !hasPermission {
+			return entity.Invitation{}, errs.NewError(errs.PermissionDenied, fmt.Sprintf("permission denied: authorization query=%v", query))
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
			return entity.Invitation{}, errs.NewError(
			    errs.PermissionDenied, 
			    fmt.Sprintf("permission denied: authorization query=%v", query))
```

File: core/service/invitation.go [hunk 5]
```diff
@@ -225,73 +243,134 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+		}
+	}
+
+	if invitation.Code != invitationCode {
+		return entity.Invitation{}, errs.NewError(errs.PermissionDenied, fmt.Sprintf("invalid invitation code: invitationID=%v, invitationCode=%v",
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
		return entity.Invitation{}, errs.NewError(
		    errs.InvalidArgument, 
		    fmt.Sprintf("invalid invitation code: invitationID=%v, invitationCode=%v",
```

File: core/service/invitation.go [hunk 6]
```diff
@@ -225,73 +243,134 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+	}
+	err = txCtx.withTransactions(false, func(tx *transaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		updateInvitationMutation := mutation.NewUpdateInvitationMutation(
+			i.dataCollector,
+			i.stateSyncer,
+			i.invitationDao,
+			i.invitationDaoV2,
+			invitation,
+		)
+
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
```

File: core/service/invitation.go [hunk 7]
```diff
@@ -225,73 +243,134 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+		internalErr := updateInvitationMutation.ExecuteV2(ct, tx)
+		if internalErr != nil {
+			return internalErr
 		}
 
-		_, err = i.teamService.AddMemberToTeam(ct, invitation.TeamID, receiverUserID)
+		rtTx.AppendMutation(updateInvitationMutation)
+
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
```

File: core/service/invitation.go [hunk 8]
```diff
@@ -225,73 +243,134 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+			if err.Code != errs.NotFound {
+				return err
+			}
+
+			teamMember := entity.TeamMember{
+				TeamID:    invitation.TeamID,
+				UserID:    receiverUserID,
+				CreatedAt: time.Now().UTC(),
+			}
+
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
```

File: core/service/invitation.go [hunk 9]
```diff
@@ -362,26 +437,11 @@ func (i Invitation) DeclineInvitation(ct context.Context, invitationID uint64, i
...
-		return internalErr
+		return errs.NewError(errs.InvalidOperation, fmt.Sprintf("invitation is revoked: invitationID=%v", invitation.ID))
 	case entity.InvitationStatusAccepted, entity.InvitationStatusDeclined:
-		internalErr := &errs.Error{
-			Code:    errs.InvalidOperation,
-			Message: fmt.Sprintf("invitation is already responded: invitationID=%v", invitation.ID),
-		}
-		i.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return internalErr
+		return errs.NewError(errs.InvalidOperation, fmt.Sprintf("invitation is already responded: invitationID=%v", invitation.ID))
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
		return errs.NewError(
		    errs.InvalidOperation, 
		    fmt.Sprintf("invitation is already responded: invitationID=%v", invitation.ID))
```

File: core/service/invitation.go [hunk 10]
```diff
@@ -362,26 +437,11 @@ func (i Invitation) DeclineInvitation(ct context.Context, invitationID uint64, i
 func (i Invitation) ensureInvitationPending(ct context.Context, invitation entity.Invitation) *errs.Error {
 	switch invitation.Status {
 	case entity.InvitationStatusExpired:
-		internalErr := &errs.Error{
-			Code:    errs.InvalidOperation,
-			Message: fmt.Sprintf("invitation is expired: invitationID=%v", invitation.ID),
-		}
-		i.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return internalErr
+		return errs.NewError(errs.InvalidOperation, fmt.Sprintf("invitation is expired: invitationID=%v", invitation.ID))
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
		return errs.NewError(
		    errs.InvalidOperation, 
		    fmt.Sprintf("invitation is expired: invitationID=%v", invitation.ID))
```

File: core/service/invitation.go [hunk 11]
```diff
@@ -362,26 +437,11 @@ func (i Invitation) DeclineInvitation(ct context.Context, invitationID uint64, i
...
 	case entity.InvitationStatusInvoked:
-		internalErr := &errs.Error{
-			Code:    errs.InvalidOperation,
-			Message: fmt.Sprintf("invitation is revoked: invitationID=%v", invitation.ID),
-		}
-		i.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return internalErr
+		return errs.NewError(errs.InvalidOperation, fmt.Sprintf("invitation is revoked: invitationID=%v", invitation.ID))
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
		return errs.NewError(
		    errs.InvalidOperation, 
		    fmt.Sprintf("invitation is revoked: invitationID=%v", invitation.ID))
```

File: core/service/invitation_test.go [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+		invitationDao,
+		invitationDaoV2,
+		teamMemberDao,
+		teamMemberDaoV2,
+		sprintParticipantDao,
+		sprintParticipantDaoV2,
+		sprintDao,
+		sprintDaoV2,
+	)
+
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
```

File: core/service/invitation_test.go [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation2.ReceiverFirstName), *(invitationFound[0].ReceiverFirstName))
+	assert.Equal(t, invitation2.ReceiverUserID, invitationFound[0].ReceiverUserID)
+	assert.Equal(t, invitation2.TeamID, invitationFound[0].TeamID)
+	assert.Equal(t, invitation2.Status, invitationFound[0].Status)
+	assert.Equal(t, invitation2.ExpireAt, invitationFound[0].ExpireAt)
+	assert.Equal(t, invitation2.CreatedAt, invitationFound[0].CreatedAt)
+	assert.Equal(t, *(invitation2.UpdatedAt), *(invitationFound[0].UpdatedAt))
+
+	filter2 := InvitationFilter{InvitationID: &invitationID2, Code: &code2}
+	invitationFound, internalErr = invitationRef.invitationService.FindInvitationsInTeam(ct, teamID1, &filter2)
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
	invitationsFound, internalErr = invitationRef.invitationService.FindInvitationsInTeam(ct, teamID1, &filter2)
```

File: core/service/invitation_test.go [hunk 14]
```diff
@@ -0,0 +1,722 @@
...
+	if !assert.Nil(t, internalErr) {
+		return
+	}
+
+	assert.Equal(t, 0, len(invitationFound))
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
	assert.Equal(t, 0, len(invitationsFound))
```

File: core/service/invitation_test.go [hunk 15]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, invitation2.ID, invitationFound[0].ID)
+	assert.Equal(t, *(invitation2.ReceiverEmail), *(invitationFound[0].ReceiverEmail))
+	assert.Equal(t, *(invitation2.ReceiverLastName), *(invitationFound[0].ReceiverLastName))
+	assert.Equal(t, *(invitation2.ReceiverFirstName), *(invitationFound[0].ReceiverFirstName))
+	assert.Equal(t, invitation2.ReceiverUserID, invitationFound[0].ReceiverUserID)
+	assert.Equal(t, invitation2.TeamID, invitationFound[0].TeamID)
+	assert.Equal(t, invitation2.Status, invitationFound[0].Status)
+	assert.Equal(t, invitation2.ExpireAt, invitationFound[0].ExpireAt)
+	assert.Equal(t, invitation2.CreatedAt, invitationFound[0].CreatedAt)
+	assert.Equal(t, *(invitation2.UpdatedAt), *(invitationFound[0].UpdatedAt))
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Can we create a helper function for this to encourage code reuse?

File: core/service/invitation_test.go [hunk 16]
```diff
@@ -0,0 +1,722 @@
...
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use the previously created helper function here

File: core/service/invitation_test.go [hunk 17]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverEmail), *(accepted.ReceiverEmail))
+	assert.Equal(t, *(invitation.ReceiverLastName), *(accepted.ReceiverLastName))
+	assert.Equal(t, *(invitation.ReceiverFirstName), *(accepted.ReceiverFirstName))
+	assert.Equal(t, requesterUserID, *(accepted.ReceiverUserID))
+	assert.Equal(t, invitation.TeamID, accepted.TeamID)
+	assert.Equal(t, entity.InvitationStatusAccepted, accepted.Status)
+	assert.Equal(t, invitation.ExpireAt, accepted.ExpireAt)
+	assert.Equal(t, invitation.Code, accepted.Code)
+	assert.Equal(t, invitation.SenderUserID, accepted.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, accepted.CreatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 18]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverLastName), *(invitationInMemory.ReceiverLastName))
+	assert.Equal(t, *(invitation.ReceiverFirstName), *(invitationInMemory.ReceiverFirstName))
+	assert.Equal(t, requesterUserID, *(invitationInMemory.ReceiverUserID))
+	assert.Equal(t, invitation.TeamID, invitationInMemory.TeamID)
+	assert.Equal(t, entity.InvitationStatusAccepted, invitationInMemory.Status)
+	assert.Equal(t, invitation.ExpireAt, invitationInMemory.ExpireAt)
+	assert.Equal(t, invitation.Code, invitationInMemory.Code)
+	assert.Equal(t, invitation.SenderUserID, invitationInMemory.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, invitationInMemory.CreatedAt)
+	assert.NotNil(t, invitationInMemory.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 19]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverLastName), *(accepted.ReceiverLastName))
+	assert.Equal(t, *(invitation.ReceiverFirstName), *(accepted.ReceiverFirstName))
+	assert.Equal(t, requesterUserID, *(accepted.ReceiverUserID))
+	assert.Equal(t, invitation.TeamID, accepted.TeamID)
+	assert.Equal(t, entity.InvitationStatusDeclined, accepted.Status)
+	assert.Equal(t, invitation.ExpireAt, accepted.ExpireAt)
+	assert.Equal(t, invitation.Code, accepted.Code)
+	assert.Equal(t, invitation.SenderUserID, accepted.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, accepted.CreatedAt)
+	assert.NotNil(t, accepted.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 20]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverLastName), *(invitationInMemory.ReceiverLastName))
+	assert.Equal(t, *(invitation.ReceiverFirstName), *(invitationInMemory.ReceiverFirstName))
+	assert.Equal(t, requesterUserID, *(invitationInMemory.ReceiverUserID))
+	assert.Equal(t, invitation.TeamID, invitationInMemory.TeamID)
+	assert.Equal(t, entity.InvitationStatusDeclined, invitationInMemory.Status)
+	assert.Equal(t, invitation.ExpireAt, invitationInMemory.ExpireAt)
+	assert.Equal(t, invitation.Code, invitationInMemory.Code)
+	assert.Equal(t, invitation.SenderUserID, invitationInMemory.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, invitationInMemory.CreatedAt)
+	assert.NotNil(t, invitationInMemory.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 21]
```diff
@@ -0,0 +1,722 @@
...
+	// verify return result
+	assert.Equal(t, *(invitation.ReceiverEmail), *(deleted.ReceiverEmail))
+	assert.Equal(t, *(invitation.ReceiverLastName), *(deleted.ReceiverLastName))
+	assert.Equal(t, *(invitation.ReceiverFirstName), *(deleted.ReceiverFirstName))
+	assert.Equal(t, invitation.TeamID, deleted.TeamID)
+	assert.Equal(t, invitation.Status, deleted.Status)
+	assert.Equal(t, invitation.ExpireAt, deleted.ExpireAt)
+	assert.Equal(t, invitation.Code, deleted.Code)
+	assert.Equal(t, invitation.SenderUserID, deleted.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, deleted.CreatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 22]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverEmail), *(invitationInMemory.ReceiverEmail))
+	assert.Equal(t, updatedLastName, *(invitationInMemory.ReceiverLastName))
+	assert.Equal(t, updatedFirstName, *(invitationInMemory.ReceiverFirstName))
+	assert.Equal(t, invitation.TeamID, invitationInMemory.TeamID)
+	assert.Equal(t, invitation.Status, invitationInMemory.Status)
+	assert.Equal(t, expiredAt, invitationInMemory.ExpireAt)
+	assert.Equal(t, invitation.Code, invitationInMemory.Code)
+	assert.Equal(t, invitation.SenderUserID, invitationInMemory.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, invitationInMemory.CreatedAt)
+	assert.NotNil(t, invitationInMemory.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 23]
```diff
@@ -0,0 +1,722 @@
...
+	assert.Equal(t, *(invitation.ReceiverEmail), *(updated.ReceiverEmail))
+	assert.Equal(t, updatedLastName, *(updated.ReceiverLastName))
+	assert.Equal(t, updatedFirstName, *(updated.ReceiverFirstName))
+	assert.Equal(t, invitation.TeamID, updated.TeamID)
+	assert.Equal(t, invitation.Status, updated.Status)
+	assert.Equal(t, expiredAt, updated.ExpireAt)
+	assert.Equal(t, invitation.Code, updated.Code)
+	assert.Equal(t, invitation.SenderUserID, updated.SenderUserID)
+	assert.Equal(t, invitation.CreatedAt, updated.CreatedAt)
+	assert.NotNil(t, updated.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 24]
```diff
@@ -0,0 +1,722 @@
...
+	}
+
+	assert.Equal(t, receiverEmail, *(invitationInMemory.ReceiverEmail))
+	assert.Equal(t, receiverLastName, *(invitationInMemory.ReceiverLastName))
+	assert.Equal(t, receiverFirstName, *(invitationInMemory.ReceiverFirstName))
+	assert.Equal(t, teamID, invitationInMemory.TeamID)
+	assert.Equal(t, entity.InvitationStatusPending, invitationInMemory.Status)
+	assert.Equal(t, expiredAt, invitationInMemory.ExpireAt)
+	assert.NotNil(t, invitationInMemory.CreatedAt)
+	assert.Nil(t, invitationInMemory.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go:395]: Let's use previously created helper function

File: core/service/invitation_test.go [hunk 25]
```diff
@@ -0,0 +1,722 @@
...
+
+	// verify return result
+	assert.Equal(t, receiverEmail, *(invitation.ReceiverEmail))
+	assert.Equal(t, receiverLastName, *(invitation.ReceiverLastName))
+	assert.Equal(t, receiverFirstName, *(invitation.ReceiverFirstName))
+	assert.Equal(t, teamID, invitation.TeamID)
+	assert.Equal(t, entity.InvitationStatusPending, invitation.Status)
+	assert.Equal(t, expiredAt, invitation.ExpireAt)
+	assert.NotNil(t, invitation.CreatedAt)
+	assert.Nil(t, invitation.UpdatedAt)
```
Comment (naturetime10) [inline core/service/invitation_test.go:380]: Let's use previously created helper function

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/daov2/sqldb/invitation.go [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+func (i Invitation) CreateInvitation(ct context.Context, tx *transaction.Transaction, invitation entity.Invitation) *errs.Error {
+	_, err := tx.SQLTx().Exec(`
+	INSERT INTO invitation
+	(
+	    id,
+		sender_user_id,
+		receiver_first_name,
+	 	receiver_last_name,
+		receiver_email,
+	    team_id,
```
Comment (naturetime10) [inline core/daov2/sqldb/invitation.go:221]:
```suggestion
	        id,
		sender_user_id,
		receiver_first_name,
	 	receiver_last_name,
		receiver_email,
	        team_id,
```

File: core/service/invitation.go [hunk 27]
```diff
@@ -159,63 +163,79 @@ func (i Invitation) CreateInvitation(ct context.Context, teamID uint64, input Cr
...
+	}
+	err = txCtx.withTransactions(false, func(tx *transaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		updateInvitationMutation := mutation.NewUpdateInvitationMutation(
+			i.logger,
+			i.stateSyncer,
+			i.invitationDao,
+			i.invitationDaoV2,
+			invitation,
+		)
+
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
```

File: core/service/invitation.go [hunk 28]
```diff
@@ -159,63 +163,79 @@ func (i Invitation) CreateInvitation(ct context.Context, teamID uint64, input Cr
...
+	}
+	err = txCtx.withTransactions(false, func(tx *transaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		deleteInvitationMutation := mutation.NewDeleteInvitationMutation(
+			i.logger,
+			i.stateSyncer,
+			i.invitationDao,
+			i.invitationDaoV2,
+			invitation,
+		)
+
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
```

File: core/service/invitation.go [hunk 29]
```diff
@@ -225,73 +245,135 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
-			i.logger.ErrorWithContext(ct, err)
-			return entity.Invitation{}, err
+			if err.Code != errs.NotFound {
+				return err
+			}
+
+			teamMember := entity.TeamMember{
+				TeamID:    invitation.TeamID,
+				UserID:    receiverUserID,
+				CreatedAt: time.Now().UTC(),
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
				CreatedAt: now,
```

File: core/service/invitation.go [hunk 30]
```diff
@@ -225,73 +245,135 @@ func (i Invitation) DeleteInvitation(ct context.Context, invitationID uint64) (e
...
+				}
+
+				return true
+			})
+
+			for _, sprint := range currAndFutureSprints {
+				participant := entity.SprintParticipant{
+					SprintID:  sprint.ID,
+					UserID:    receiverUserID,
+					CreatedAt: time.Now().UTC(),
```
Comment (naturetime10) [inline core/service/invitation.go]:
```suggestion
					CreatedAt: now,
```

File: core/service/invitation_test.go [hunk 31]
```diff
@@ -0,0 +1,724 @@
...
+	if one.ExpireAt != other.ExpireAt {
+		return false
+	}
+
+	if one.CreatedAt != other.CreatedAt {
+		return false
+	}
+
+	if one.ReceiverUserID == nil || other.ReceiverUserID == nil {
+		if !(one.ReceiverUserID == nil && other.ReceiverUserID == nil) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if one.ReceiverUserID != nil || other.ReceiverUserID != nil {
```

File: core/service/invitation_test.go [hunk 32]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	} else {
+		if *(one.ReceiverUserID) != *(other.ReceiverUserID) {
+			return false
+		}
+	}
+
+	if one.ReceiverLastName == nil || other.ReceiverLastName == nil {
+		if !(one.ReceiverLastName == nil && other.ReceiverLastName == nil) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if one.ReceiverLastName != nil || other.ReceiverLastName != nil) {
```

File: core/service/invitation_test.go [hunk 33]
```diff
@@ -0,0 +1,724 @@
...
+	if one.CreatedAt != other.CreatedAt {
+		return false
+	}
+
+	if one.ReceiverUserID == nil || other.ReceiverUserID == nil {
+		if !(one.ReceiverUserID == nil && other.ReceiverUserID == nil) {
+			return false
+		}
+	} else {
+		if *(one.ReceiverUserID) != *(other.ReceiverUserID) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if *one.ReceiverUserID) != *other.ReceiverUserID {
```

File: core/service/invitation_test.go [hunk 34]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	}
+
+	if one.ReceiverLastName == nil || other.ReceiverLastName == nil {
+		if !(one.ReceiverLastName == nil && other.ReceiverLastName == nil) {
+			return false
+		}
+	} else {
+		if *(one.ReceiverLastName) != *(other.ReceiverLastName) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if *one.ReceiverLastName) != *other.ReceiverLastName) {
```

File: core/service/invitation_test.go [hunk 35]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	}
+
+	if one.ReceiverFirstName == nil || other.ReceiverFirstName == nil {
+		if !(one.ReceiverFirstName == nil && other.ReceiverFirstName == nil) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if one.ReceiverFirstName != nil || other.ReceiverFirstName != nil) {
```

File: core/service/invitation_test.go [hunk 36]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	} else {
+		if *(one.ReceiverFirstName) != *(other.ReceiverFirstName) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if *one.ReceiverFirstName != *other.ReceiverFirstName {
```

File: core/service/invitation_test.go [hunk 37]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	}
+
+	if one.ReceiverEmail == nil || other.ReceiverEmail == nil {
+		if !(one.ReceiverEmail == nil && other.ReceiverEmail == nil) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if one.ReceiverEmail != nil || other.ReceiverEmail != nil {
```

File: core/service/invitation_test.go [hunk 38]
```diff
@@ -0,0 +1,724 @@
...
+			return false
+		}
+	} else {
+		if *(one.ReceiverEmail) != *(other.ReceiverEmail) {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
		if *one.ReceiverEmail != *other.ReceiverEmail {
```

File: core/service/invitation_test.go [hunk 39]
```diff
@@ -0,0 +1,724 @@
...
+		if *(one.ReceiverUserID) != *(other.ReceiverUserID) {
+			return false
+		}
+	}
+
+	if one.ReceiverLastName == nil || other.ReceiverLastName == nil {
+		if !(one.ReceiverLastName == nil && other.ReceiverLastName == nil) {
+			return false
+		}
+	} else {
```
Comment (naturetime10) [inline core/service/invitation_test.go]: Maybe merge all 
```
else {
    if {}
}

to
else if {}
```
?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/invitation_test.go [hunk 40]
```diff
@@ -0,0 +1,724 @@
...
+	invitationInMemory, err := invitationRef.invitationDaoV2.FindInvitationByIDWithTx(ct, tx, invitation.ID)
+	if !assert.Nil(t, err) {
+		return
+	}
+
+	assert.Equal(t, true, isInvitationEqual(invitation, invitationInMemory))
+}
+
+// helper
+func isInvitationEqual(one entity.Invitation, other entity.Invitation) bool {
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
func areInvitationsEqual(one entity.Invitation, other entity.Invitation) bool {
```

File: core/service/invitation_test.go [hunk 41]
```diff
@@ -0,0 +1,724 @@
...
+
+	filter1 := InvitationFilter{InvitationID: &invitationID2, Code: &code1}
+	invitationsFound, internalErr := invitationRef.invitationService.FindInvitations(ct, &filter1)
+	if !assert.Nil(t, internalErr) {
+		return
+	}
+
+	// verify return result
+	assert.Equal(t, 1, len(invitationsFound))
+	assert.Equal(t, true, isInvitationEqual(invitation2, invitationsFound[0]))
```
Comment (naturetime10) [inline core/service/invitation_test.go]: assert.True

File: core/service/invitation_test.go [hunk 42]
```diff
@@ -0,0 +1,724 @@
...
+	invitation.ReceiverLastName = &updatedLastName
+	assert.Equal(t, true, isInvitationEqual(invitation, updated))
+
+	// verify in-memory DB
+	invitationInMemory, err := invitationRef.invitationDaoV2.FindInvitationByIDWithTx(ct, tx, invitation.ID)
+	if !assert.Nil(t, err) {
+		return
+	}
+
+	assert.Equal(t, true, isInvitationEqual(invitation, invitationInMemory))
```
Comment (naturetime10) [inline core/service/invitation_test.go]: assert.True

File: core/service/invitation_test.go [hunk 43]
```diff
@@ -0,0 +1,724 @@
...
+
+	// verify return result
+	assert.Equal(t, true, isInvitationEqual(invitation, deleted))
+
+	// verify in-memory DB
+	_, err = invitationRef.invitationDaoV2.FindInvitationByIDWithTx(ct, tx, invitation.ID)
+	if !assert.NotNil(t, err) {
+		return
+	}
+	assert.Equal(t, errs.NotFound, err.Code)
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
	}
	
	assert.Equal(t, errs.NotFound, err.Code)
```

File: core/service/invitation_test.go [hunk 44]
```diff
@@ -0,0 +1,724 @@
...
+		return
+	}
+
+	deleted, err := invitationRef.invitationService.DeleteInvitation(ct, invitation.ID)
+	if !assert.Nil(t, err) {
+		return
+	}
+
+	// verify return result
+	assert.Equal(t, true, isInvitationEqual(invitation, deleted))
```
Comment (naturetime10) [inline core/service/invitation_test.go]: assert.True

File: core/service/invitation_test.go [hunk 45]
```diff
@@ -0,0 +1,724 @@
...
+	invitation.ReceiverUserID = &requesterUserID
+	assert.Equal(t, true, isInvitationEqual(invitation, accepted))
+
+	// verify in-memory DB
+	invitationInMemory, err := invitationRef.invitationDaoV2.FindInvitationByIDWithTx(ct, tx, invitation.ID)
+	if !assert.Nil(t, err) {
+		return
+	}
+
+	assert.Equal(t, true, isInvitationEqual(invitation, invitationInMemory))
```
Comment (naturetime10) [inline core/service/invitation_test.go]: assert.True

File: core/service/invitation_test.go [hunk 46]
```diff
@@ -0,0 +1,724 @@
...
+}
+
+// helper
```
Comment (naturetime10) [inline core/service/invitation_test.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/daov2/sqldb/invitation.go (hunk 26 above)
Comment (Junoth) [reply core/daov2/sqldb/invitation.go:221]: This seems to be github format issue. I tried goland and vscode in my local machine and this looked good
Screen Shot 2023-04-06 at 11 09 57 PM

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**